sdk/python/kfp/_local_client.py

```diff
@@ -0,0 +1,441 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import datetime
+import json
+import logging
+import re
+import subprocess
+from typing import Any, Callable, Dict, List, Mapping, Tuple, Union, cast
+
+import kfp
+
+
+class _Dag:
+    def __init__(self, nodes):
+        self.graph = {node: [] for node in nodes}
```

Comment:
It looks like `graph` and `reverse_graph` are always in-sync, right?

If so shall we consider ensuring this is always true by leveraging `property` and `setter`?

Reply:
Yes, `graph` and `reverse_graph` are always in sync.
I have changed their visibility to private. Is that necessary to use property? How?

Reply:
Even though they are changed to private, caller can still change its value because Python does not have much visibility control.

What I meant is something like the following:
```python
@property
def graph(self):
  return self._graph

@property
def reverse_graph(self):
  return self._reverse_graph
```
This way, the only accessible API to mutate their value is through `add_edge` method, where the sync is guaranteed.